gtda/diagrams/_metrics.py

```diff
@@ -284,34 +343,60 @@ def wasserstein_amplitudes(diagrams, p=2., **kwargs):
 
 
 def betti_amplitudes(diagrams, sampling, step_size, p=2., **kwargs):
+    step_size_factor = step_size ** (1 / p)
     bcs = betti_curves(diagrams, sampling)
-    return (step_size ** (1 / p)) * np.linalg.norm(bcs, axis=1, ord=p)
+    amplitudes = np.linalg.norm(bcs, axis=1, ord=p)
+    amplitudes *= step_size_factor
+    return amplitudes
 
 
-def landscape_amplitudes(diagrams, sampling, step_size, p=2., n_layers=1,
-                         **kwargs):
+def landscape_amplitudes(
+        diagrams, sampling, step_size, p=2., n_layers=1, **kwargs
+        ):
+    step_size_factor = step_size ** (1 / p)
     ls = landscapes(diagrams, sampling, n_layers).\
         reshape(len(diagrams), -1)
-    return (step_size ** (1 / p)) * np.linalg.norm(ls, axis=1, ord=p)
-
-
-def heat_amplitudes(diagrams, sampling, step_size, sigma=1., p=2., **kwargs):
-    heat = heats(diagrams, sampling, step_size, sigma)
-    return np.linalg.norm(heat, axis=(1, 2), ord=p)
+    amplitudes = np.linalg.norm(ls, axis=1, ord=p)
+    amplitudes *= step_size_factor
+    return amplitudes
 
 
-def persistence_image_amplitudes(diagrams, sampling, step_size,
-                                 weight_function=lambda x: x, sigma=1., p=2.,
-                                 **kwargs):
-    persistence_image = persistence_images(diagrams, sampling, step_size,
-                                           weight_function, sigma)
-    return np.linalg.norm(persistence_image, axis=(1, 2), ord=p)
-
-
-def silhouette_amplitudes(diagrams, sampling, step_size, power=2., p=2.,
-                          **kwargs):
-    sht = silhouettes(diagrams, sampling, power)
-    return (step_size ** (1 / p)) * np.linalg.norm(sht, axis=1, ord=p)
+def heat_amplitudes(diagrams, sampling, step_size, sigma=0.1, p=2., **kwargs):
+    # WARNING: `heats` modifies `diagrams` in place
+    step_size_factor = step_size ** (2 / p)
+    heats_ = heats(diagrams, sampling, step_size, sigma).\
+        reshape(len(diagrams), -1)
+    amplitudes = np.linalg.norm(heats_, axis=1, ord=p)
+    amplitudes *= step_size_factor
+    return amplitudes
+
+
+def persistence_image_amplitudes(
+        diagrams, sampling, step_size, sigma=0.1, weight_function=np.ones_like,
+        p=2., **kwargs
+        ):
+    # For persistence images, `sampling` is a tall matrix with two columns
+    # (the first for birth and the second for persistence), and `step_size` is
+    # a 2d array
+    weights = weight_function(sampling[:, 1])
+    step_sizes_factor = np.product(step_size) ** (1 / p)
+    # WARNING: `persistence_images` modifies `diagrams` in place
+    persistence_images_ = persistence_images(
+        diagrams, sampling, step_size, sigma, weights
+        ).reshape(len(diagrams), -1)
```

Comment:
I like how you consistently reshape into a 2d array - super clear where the norm is then.

Reply:
Right. Not only that, the current approach was also incorrect! It was computing the 2-norm of matrices, which is the absolute value of the largest singular value. But we wanted the Frobenius norm in that case...

Reply:
And it failed when p was a float for that reason...